marlowe/src/Language/Marlowe/Core/V1/Semantics/Next.hs

```diff
@@ -0,0 +1,175 @@
+
+newtype CaseIndex = CaseIndex {unIndex ::Integer}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+    deriving (FromJSON,ToJSON)
+
+-- | Next Ranges of Inputs that can be applied to a given contract
+data NextGeneralizedInput
+  = CanDeposit CaseIndex Party AccountId Token Integer -- 1 possibility > 1 action
+  | CanChoose  CaseIndex ChoiceId [Bound] -- 1 possibility > many different actions
+  | CanNotify  CaseIndex -- 1 possibility > 1 action
+  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+  deriving anyclass (Pretty)
+
+
+next :: Environment ->  State  ->  Contract -> Either AmbiguousIntervalProvided Next
+next  _ _ Close         = Right . CanAdvance $ []
+next  _ _ (When [] _ _) = Right CanNotAdvance   -- Contract Suspended in a time out
+next environment state (When (x:xs) _ _)
+  = Right
+    . maybe
+          CanNotAdvance -- Contract Suspended (only falsified Notifies evaluated in the When)
+        ( CanAdvance . keepFirstNotifyTrueAndRemoveFollowings  )
+    . nextGeneralizedInputsMaybe environment state
+    $ x :| xs
+next environment state  contract  = uncurry (next environment) =<< reduceContract environment state contract
+
+
+keepFirstNotifyTrueAndRemoveFollowings ::  NonEmpty NextGeneralizedInput -> [NextGeneralizedInput]
+keepFirstNotifyTrueAndRemoveFollowings
+  = keepFirstNotifyTrueAndRemoveFollowings' False . NEL.toList
+ where
+  keepFirstNotifyTrueAndRemoveFollowings' :: Bool -> [NextGeneralizedInput] -> [NextGeneralizedInput]
+  keepFirstNotifyTrueAndRemoveFollowings' _ [] =  []
+  keepFirstNotifyTrueAndRemoveFollowings' False (x@CanNotify {} : xs) =  keepFirstNotifyTrueAndRemoveFollowings' True (x:xs)
+  keepFirstNotifyTrueAndRemoveFollowings' True (CanNotify {} : xs) =  keepFirstNotifyTrueAndRemoveFollowings' True xs
+  keepFirstNotifyTrueAndRemoveFollowings' _ xs =  keepFirstNotifyTrueAndRemoveFollowings' True xs
+
+
+nextGeneralizedInputsMaybe :: Environment -> State -> NonEmpty (Case Contract) -> Maybe (NonEmpty NextGeneralizedInput)
+nextGeneralizedInputsMaybe environment state
+  = nonEmpty
+  . catMaybes
+  . (uncurry (nextGeneralizedInputMaybe environment state)  <$>)
+  . caseIndexed
+
+nextGeneralizedInputMaybe :: Environment -> State -> CaseIndex -> Case Contract -> Maybe NextGeneralizedInput
+nextGeneralizedInputMaybe environment state caseIndex
+  = \case
+      (Deposit accountId party token value)
+        -> Just $ CanDeposit
+                caseIndex
+                accountId
+                party
+                token
+                (evalValue environment state value )
+      (Choice choiceId bounds) -> Just $ CanChoose caseIndex choiceId bounds
+      (Notify observation) | evalObservation environment state observation -> Just $ CanNotify caseIndex
+      (Notify _) -> Nothing
+  . \case
+      (Case action _)           -> action
+      (MerkleizedCase action _) -> action
+
+caseIndexed :: NonEmpty a -> [(CaseIndex,a)]
```

Comment:
Huh, so you had the case position separate initially, if you don't combine it with `NextGeneralizedInput` the whole `nextGeneralizedInputMaybe` gets simpler